docs/core/extensions/options-validation-generator.md

```diff
@@ -0,0 +1,262 @@
+---
+title: Compile-time options validation source generation
+author: tarekgh
+description: Learn how to utilize compile-time source generation for validating options.
+ms.author: tarekms
+ms.date: 10/23/2023
+---
+
+# Compile-time options validation source generation
+
+In the [options pattern](options.md), various methods for validating options are presented. These methods include using data annotation attributes or employing a custom validator. Data annotation attributes are validated at runtime and can incur performance costs. This article demonstrates how to utilize the options validation source generator to produce optimized validation code during compile time.
+
+## Automatic IValidateOptions implementation generation
+
+In the [options pattern](options.md), we've illustrated how to implement the IValidateOptions interface for validating options. The options validation source generator can automatically create the IValidateOptions interface implementation by leveraging data annotation attributes on the options class.
+
+Let's take previous annotation attributes example which mentioned in [options pattern](options.md) to convert it to use the options validation source generator.
+
+Consider the following *appsettings.json* file:
+
+```json
+{
+  "MyCustomSettingsSection": {
+    "SiteTitle": "Amazing docs from Awesome people!",
+    "Scale": 10,
+    "VerbosityLevel": 32
+  }
+}
+```
+
+The following class binds to the `"MyCustomSettingsSection"` configuration section and applies a couple of `DataAnnotations` rules:
+
+:::code language="csharp" source="snippets/configuration/console-json/SettingsOptions.cs":::
+
+In the preceding `SettingsOptions` class, the `ConfigurationSectionName` property contains the name of the configuration section to bind to. In this scenario, the options object provides the name of its configuration section. Then we call the following code to bind the configuration section to the options object and validate the data annotations:
+
+```csharp
+builder.Services
+    .AddOptions<SettingsOptions>()
+    .Bind(Configuration.GetSection(SettingsOptions.ConfigurationSectionName))
+    .ValidateDataAnnotations();
+```
+
+While this approach works effectively, it can introduce performance implications because validation occurs at runtime via reflection. Additionally, reliance on reflection can lead to issues, particularly when using AOT compilation. Leveraging compile-time source generation for options validation allows you to generate performance-optimized validation code and eliminates the need for reflection, resulting in smoother AOT-compatible app building. The following code demonstrates how to use the options validation source generator:
```

Comment:
```suggestion
While this approach is effective, it can affect performance because validation occurs at run time via reflection. Additionally, reliance on reflection can lead to issues, particularly when using AOT compilation. By leveraging compile-time source generation for options validation, you can generate performance-optimized validation code and eliminate the need for reflection, resulting in smoother AOT-compatible app building. The following code demonstrates how to use the options validation source generator:
```